### PR TITLE
Service Workers

### PR DESCRIPTION
**Pull request by artpi**
# WHAT?
Yes, we are experimenting with service workers, although we encountered some problems.

# How do I run this branch?
Because service workers demand https, you need to run Chrome with special flags to run this.
```
/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --unsafely-treat-insecure-origin-as-secure=http://calypso.localhost:3000 --user-data-dir=/tmp/service-workers-chrome-user-dir
```
# What now?
Please, discuss.

CC @gwwar @rralian : I have cherry-picked these commits from our meetup branch and rebased.

### CODE REVIEW

**Comment by artpi** [comment]
Mostly, the problems we have encountered revolve around:

- development flags
- killing service workers (they are zombie-like)
- updating cache

**Comment by ehg** [comment]
> Because service workers demand https, you need to run Chrome with special flags to run this.

http://localhost is [whitelisted](https://www.chromium.org/blink/serviceworker/service-worker-faq) in Chrome, so if we run Calypso on that address, we won't need to supply the special flags.

**Comment by ehg** [comment]
Have you seen https://github.com/NekR/offline-plugin ?

File: server/pages/serviceWorker.js [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+				// CORS (http://en.wikipedia.org/wiki/Cross-origin_resource_sharing).
+				// In this example, www.chromium.org doesn't support CORS, and the fetch()
+				// would fail if the default mode of 'cors' was used for the fetch() request.
+				// The drawback of hardcoding {mode: 'no-cors'} is that the response from all
+				// cross-origin hosts will always be opaque
+				// (https://slightlyoff.github.io/ServiceWorker/spec/service_worker/index.html#cross-origin-resources)
+				// and it is not possible to determine whether an opaque response represents a success or failure
+				// (https://github.com/whatwg/fetch/issues/14).
+				return fetch( new Request( url, { mode: 'no-cors' } ) ).then( function( response ) {
+					if ( response.status >= 400 ) {
```
Comment (delapuente) [inline server/pages/serviceWorker.js]: With `mode: 'no-cors' ` you will receive opaque response. The [status for these responses is always 0](https://fetch.spec.whatwg.org/#concept-filtered-response-opaque) cause it's intended you can extract no knowledge about them at all.

File: server/pages/serviceWorker.js [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+// If at any point you want to force pages that use this service worker to start using a fresh
+// cache, then increment the CACHE_VERSION value. It will kick off the service worker update
+// flow and the old cache(s) will be purged as part of the activate event handler when the
+// updated service worker is activated.
+var CACHE_VERSION = 1;
+var CURRENT_CACHES = {
+	prefetch: 'calypso-prefetch-cache-v' + CACHE_VERSION
+};
+
+self.addEventListener('install', function( event ) {
```
Comment (delapuente) [inline server/pages/serviceWorker.js]: Now you could `self.skipWaiting()` to not wait until other clients being controlled by other sw close. This, in combination with `self.clients.claim()` during activation leads to controlling the clients ASAP.

See more about the life-cycle here: https://github.com/delapuente/service-workers-101/

File: server/pages/serviceWorker.js [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+				} )
+			);
+		} )
+	);
+} );
+
+self.addEventListener( 'fetch', function( event ) {
+	console.log( 'Handling fetch event for', event.request.url );
+
+	event.respondWith(
```
Comment (delapuente) [inline server/pages/serviceWorker.js]: This is a simple cache-first policy. The more complex policy of cache-first-with-background-update could be implemented as follows:

```js
caches.match( event.request ).then( function( response ) {
			var fetchFromNetwork = fetch( event.request.clone()).then( function( response ) {
				console.log( 'Response from network is:', response );
				return caches.open().then(function(cache) {
					return cache.put(event.request, response.clone()).then(function() { return response.clone(); });
				});
			} ).catch( function( error ) {
				console.error( 'Fetching failed:', error );
				throw error;
			} );
			return response ? Promise.resolve(response.clone()) || fetchFromNetwork;
		} )
```

**Comment by marco-c** [comment]
If you intend to use service workers only for offlining static assets, you can take a look at [oghliner](https://github.com/mozilla/oghliner).
It's very easy to use and it generates a service worker for you.

**Comment by lancewillett** [comment]
Closing this one as no movement in several months — feel free to re-open if it's actionable and can be pushed in.

**Pull request closed without merging**